pkg/dataobj/planner/physical/plan.go

```diff
@@ -0,0 +1,297 @@
+	Traversable
+	// ID returns a string that uniquely identifies a node in the plan
+	ID() string
+	// Type returns the node type
+	Type() NodeType
+	// isNode is a marker interface to denote a node, and only allows it to be
+	// implemented within this package
+	isNode()
+}
+
+var _ Node = (*DataObjScan)(nil)
+var _ Node = (*SortMerge)(nil)
+var _ Node = (*Projection)(nil)
+var _ Node = (*Limit)(nil)
+var _ Node = (*Filter)(nil)
+
+func (*DataObjScan) isNode() {}
+func (*SortMerge) isNode()   {}
+func (*Projection) isNode()  {}
+func (*Limit) isNode()       {}
+func (*Filter) isNode()      {}
+
+// Edge is a directed connection (parent-child relation) between a two nodes.
+type Edge struct {
+	Parent, Child Node
+}
+
+type WalkOrder uint8
+
+const (
+	PreOrderWalk  WalkOrder = iota // Pre-order: Process the current vertex before visiting any of its children.
+	PostOrderWalk                  // Post-order: Process the current vertex after visiting all of its children.
+)
+
+type nodeSet map[Node]struct{}
+
+func (s nodeSet) add(node Node) {
+	if node == nil {
+		return
+	}
+	s[node] = struct{}{}
+}
+
+func (s nodeSet) remove(node Node) {
+	if s.contains(node) {
+		delete(s, node)
+	}
+}
+
+func (s nodeSet) contains(node Node) bool {
+	if node == nil {
+		return false
+	}
+	_, ok := s[node]
+	return ok
+}
+
+func (s nodeSet) sorted() []Node {
+	nodes := make([]Node, 0, len(s))
+	for node := range s {
+		nodes = append(nodes, node)
+	}
+	slices.SortFunc(nodes, func(a, b Node) int {
+		if a.ID() > b.ID() {
+			return 1
+		}
+		if a.ID() < b.ID() {
+			return -1
+		}
+		return 0
+	})
+	return nodes
+}
+
+// Plan represents a physical execution plan as a directed acyclic graph (DAG).
+// It maintains the relationships between nodes, tracking parent-child connections
+// and providing methods for graph traversal and manipulation.
+//
+// The plan structure supports operations like adding nodes and edges,
+// retrieving nodes by ID, retrieving parents and children of nodes, and
+// walking the graph in different orders using the depth-first-search algorithm.
+type Plan struct {
+	// nodesByID maps node IDs to their corresponding Node instances for quick lookups
+	nodesByID map[string]Node
+	// nodes is a set containing all nodes in the plan
+	nodes nodeSet
+	// parents maps each node to a set of its parent nodes in the execution graph
+	parents map[Node]nodeSet
+	// children maps each node to a set of its child nodes in the execution graph
+	children map[Node]nodeSet
+}
+
+func (p *Plan) init() {
+	if p.nodesByID == nil {
+		p.nodesByID = make(map[string]Node)
+	}
+	if p.nodes == nil {
+		p.nodes = make(nodeSet)
+	}
+	if p.parents == nil {
+		p.parents = make(map[Node]nodeSet)
+	}
+	if p.children == nil {
+		p.children = make(map[Node]nodeSet)
+	}
+}
+
+// AddNode adds a new node to the plan if it doesn't already exist. For
+// convenience, the function returns the input node without modification.
+func (p *Plan) AddNode(n Node) Node {
```

Comment:
Should these methods be publicly exposed? Do we want downstream packages to be able to modify the DAG?